crates/ruff/src/rules/pylint/rules/unexpected_special_method_signature.rs

```diff
@@ -0,0 +1,132 @@
+use std::ops::RangeInclusive;
+
+use rustpython_parser::ast::{Arguments, Expr, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::identifier_range;
+use ruff_python_ast::source_code::Locator;
+use ruff_python_semantic::analyze::visibility::is_staticmethod;
+use ruff_python_semantic::scope::ScopeKind;
+
+use crate::checkers::ast::Checker;
+
+#[violation]
+pub struct UnexpectedSpecialMethodSignature {
+    pub method_name: String,
+    pub expected_params: std::ops::RangeInclusive<usize>,
+    pub actual_params: usize,
+}
+
+impl Violation for UnexpectedSpecialMethodSignature {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let verb = if self.actual_params > 1 {
+            "were"
+        } else {
+            "was"
+        };
+        if self.expected_params.end() - self.expected_params.start() > 1 {
+            format!(
+                "The special method '{}' expects between {} and {} param(s), {} {} given",
+                self.method_name,
+                self.expected_params.start(),
+                self.expected_params.end(),
+                self.actual_params,
+                verb,
+            )
+        } else {
+            format!(
+                "The special method '{}' expects {} param(s), {} {} given",
+                self.method_name,
+                self.expected_params.start(),
+                self.actual_params,
+                verb,
+            )
+        }
+    }
+}
+
+fn expected_params(name: &str, is_staticmethod: bool) -> Option<RangeInclusive<usize>> {
```

Comment:
Rather than modeling each case as `RangeInclusive`, and using `self.expected_params.end() - self.expected_params.start() > 1`-like checks to determine whether a branch should be treated as a range or not, what if we introduced an enum here to represent the two cases? E.g.:

```rust
enum ExpectedParameters {
  Fixed(usize),
  // Or RangeInclusive<usize>
  Range(usize, usize),
}
```

(Names etc. are of course flexible, just illustrating the concept.)

I think that would make some of the case-handling below a bit clearer by way of being more explicit.

Reply:
Done.

I used the new enum in the violation so it has to be public :/. If you have a preference `expected_params` can be rendered into a string for formatting before creating the violation, which would allow `ExpectedParams` to be private again.

Reply:
All good, we follow this pattern after and the enums end up public. No worries.